File: QuasselDroidNG/src/main/java/de/kuschku/quasseldroid_ng/QuasselService.java
package de.kuschku.quasseldroid_ng;

import android.app.Service;
import android.content.Intent;
import android.os.Binder;
import android.os.IBinder;

import de.kuschku.libquassel.BusProvider;
import de.kuschku.quasseldroid_ng.util.ServerAddress;

public class QuasselService extends Service {
    private final IBinder binder = new LocalBinder();

    private ClientBackgroundThread bgThread;

    @Override
    public IBinder onBind(Intent intent) {
        return binder;
    }

    @Override
    public void onCreate() {
        super.onCreate();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        return START_STICKY;
    }

    public class LocalBinder extends Binder {
        public void startBackgroundThread(BusProvider provider, ServerAddress address) {
            bgThread = new ClientBackgroundThread(provider, address);
            new Thread(bgThread).start();
        }

        public ClientBackgroundThread getBackgroundThread() {
            return bgThread;
        }

        public void stopBackgroundThread() {
            if (bgThread != null) bgThread.close();
            bgThread = null;
        }
    }
}

File: app/src/main/kotlin/de/justjanne/quasseldroid/service/QuasselService.kt
package de.justjanne.quasseldroid.service

import android.app.Service
import android.content.Context
import android.content.Intent
import android.os.IBinder
import android.util.Log
import androidx.room.Room
import de.justjanne.quasseldroid.persistence.AppDatabase
import java.net.InetSocketAddress

class QuasselService : Service() {
  private var runner: QuasselRunner? = null

  private val database: AppDatabase by lazy {
    Room.databaseBuilder(
      this.applicationContext,
      AppDatabase::class.java,
      "app"
    ).build()
  }

  private fun newRunner(intent: Intent): QuasselRunner {
    Log.w("QuasselService", "Creating new quassel runner")
    val address = InetSocketAddress.createUnresolved(
      requireNotNull(intent.getStringExtra("host")) {
        "Required argument 'host' missing"
      },
      intent.getIntExtra("port", 4242),
    )
    val auth = Pair(
      requireNotNull(intent.getStringExtra("username")) {
        "Required argument 'username' missing"
      },
      requireNotNull(intent.getStringExtra("password")) {
        "Required argument 'password' missing"
      },
    )
    return QuasselRunner(address, auth, database)
  }

  override fun onCreate() {
    Log.d("QuasselService", "Service created")
  }

  override fun onDestroy() {
    runner?.close()
    super.onDestroy()
  }

  override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
    Log.d("QuasselService", "Start Command received")
    super.onStartCommand(intent, flags, startId)
    if (intent != null && this.runner == null) {
      this.runner = newRunner(intent)
    } else if (this.runner == null) {
      Log.e("QuasselService", "Could not start runner, intent missing")
    }
    return START_STICKY
  }

  override fun onBind(intent: Intent): IBinder {
    Log.d("QuasselService", "Binding")
    return QuasselBinder(this.runner ?: newRunner(intent).also { runner = it })
  }

  companion object {
    fun Context.quasselService(
      address: InetSocketAddress,
      auth: Pair<String, String>
    ) = quasselService().apply {
      putExtra("host", address.hostString)
      putExtra("port", address.port)

      val (username, password) = auth
      putExtra("username", username)
      putExtra("password", password)
    }

    fun Context.quasselService() = Intent(applicationContext, QuasselService::class.java)
  }
}
